feat: save chart in dashboard

ZeRego opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->
Relates to: #6133 
Closes: #6227 <!-- reference the related issue e.g. #150 -->

### Description:
<!-- Add a description of the changes proposed in the pull request. -->

<!-- Even better add a screenshot / gif / loom -->

Changes:
- amend knex table typing
- update chart table to have a dashboard uuid column and space id column be nullable


Table before:
```
create table saved_queries
(
    saved_query_id                    integer generated always as identity
        primary key,
    saved_query_uuid                  uuid         default uuid_generate_v4() not null
        constraint saved_queries_saved_query_uuid_unique
            unique,
    created_at                        timestamp    default CURRENT_TIMESTAMP  not null,
    name                              text                                    not null,
    space_id                          integer                                 not null
        constraint saved_queries_space_id_foreign
            references spaces
            on delete cascade,
    description                       text,
    last_version_chart_kind           varchar(255) default 'vertical_bar'::character varying,
    last_version_updated_at           timestamp    default CURRENT_TIMESTAMP,
    last_version_updated_by_user_uuid uuid
        constraint saved_queries_last_version_updated_by_user_uuid_foreign
            references users (user_uuid)
            on delete set null
);
```

Table now:

```
create table saved_queries
(
    saved_query_id                    integer generated always as identity
        primary key,
    saved_query_uuid                  uuid         default uuid_generate_v4() not null
        constraint saved_queries_saved_query_uuid_unique
            unique,
    created_at                        timestamp    default CURRENT_TIMESTAMP  not null,
    name                              text                                    not null,
    space_id                          integer
        constraint saved_queries_space_id_foreign
            references spaces
            on delete cascade,
    description                       text,
    last_version_chart_kind           varchar(255) default 'vertical_bar'::character varying,
    last_version_updated_at           timestamp    default CURRENT_TIMESTAMP,
    last_version_updated_by_user_uuid uuid
        constraint saved_queries_last_version_updated_by_user_uuid_foreign
            references users (user_uuid)
            on delete set null,
    dashboard_uuid                    uuid
        constraint saved_queries_dashboard_uuid_foreign
            references dashboards (dashboard_uuid)
            on delete cascade
);

```

E2e test passing:

<img width="592" alt="Screenshot 2023-07-18 at 15 43 46" src="https://github.com/lightdash/lightdash/assets/9117144/7b124869-00aa-473e-8fdd-1aa2caf628f9">


Preview:

https://github.com/lightdash/lightdash/assets/9117144/7f4329e9-619c-4147-89de-3fa603b728ef

ZeRego commented on packages/backend/src/models/ProjectModel/ProjectModel.ts:
```diff
@@ -875,11 +875,17 @@ export class ProjectModel {
                                       saved_query_id: undefined,
                                       saved_query_uuid: undefined,
                                       space_id: getNewSpace(d.space_id),
+                                      dashboard_uuid: null,
```
We are going to support preview content in a following PR

rephus commented on packages/backend/src/database/migrations/20230718093901_add_dashbord_uuid_to_chart_table.ts:
```diff
@@ -0,0 +1,21 @@
+    await knex.schema.alterTable('saved_queries', (t) => {
+        t.dropColumns('dashboard_uuid');
+        // make space id non nullable
+        t.integer('space_id').notNullable().alter({ alterType: false });
```
we probably need to remove all the saved queries with space_id null to apply this

ZeRego replied:
good catch 👍

rephus commented on packages/backend/src/models/DashboardModel/DashboardModel.ts:
```diff
@@ -98,6 +98,7 @@ export class DashboardModel {
 
     private static async createVersion(
         trx: Transaction,
+        dashboardUuid: string,
```
is this not optional ? `dashboardUuid: string | undefined`

ZeRego replied:
Even though it might not be used, we don't know at this point, so it is mandatory. Note this is a method to create a dashboard version, not a chart version.

rephus commented:
> This branch has not been deployed

Can we get this in render ?

ZeRego commented on packages/backend/package.json:
```diff
@@ -116,7 +116,7 @@
         "migrate-production": "knex migrate:latest --knexfile dist/database/knexfile.js",
         "seed": "knex seed:run --knexfile src/database/knexfile.ts",
         "seed-production": "knex seed:run --knexfile dist/database/knexfile.js",
-        "rollback-last": "knex migrate:rollback --knexfile src/database/knexfile.ts",
```
`knex migrate:rollback` is to rollback the last batch of migrations. eg: If you migrated everything in 1 go, it would rollback everything. 

We just want to rollback the last migration script.

rephus commented:
Do we have this API dashboard twice because we run e2e tests twice ? 

![image](https://github.com/lightdash/lightdash/assets/1983672/b9e0c519-11aa-47d7-a6d2-e24f70b816a7)

perhaps we want to delete it "at the beggining" of a new e2e test

rephus commented:
I can't edit the title or the content of a chart in dashboard

rephus commented:
I haven't tested scheduled delivieres for these charts, but I can see they can be previewed on /minimal, so I assume they work

rephus commented:
Interesting behaviour on these chart in dashboard options: 

![image](https://github.com/lightdash/lightdash/assets/1983672/a257177c-2e96-4490-ba42-ad8eeb82cacf)

- save chart as: creates a new chat that doens't belong to a dashboard
- add to dashboard: you can add this chart to another dashboard, but I guess this will be deleted as well if the original dashboard gets deleted. 
- move to space: errors (tries to update the chart, see previous comment)

rephus reviewed (changes requested):
I think the only thing we need to fix is the ability to update chart in dashboards. Everything else seems ok, or not a blocker.

ZeRego commented:
@rephus 
> Do we have this API dashboard twice because we run e2e tests twice ?
> perhaps we want to delete it "at the beggining" of a new e2e test

We already delete these dashboards and charts at the beginning of the test. This may happen when the test retries.

> I can't edit the title or the content of a chart in dashboard

Fixed chart summary query and added test.

<img width="594" alt="Screenshot 2023-07-21 at 10 22 18" src="https://github.com/lightdash/lightdash/assets/9117144/a483f754-1035-4964-83c2-0b38b65d2b59">


> Interesting behaviour on these chart in dashboard options:

We have tickets in the milestone to disable some of these options. Shouldn't be an issue while the feature is under a feature flag.

rephus reviewed (approved):
Everything works now. I'm happy about your comments.

Just 1 minor bug, we can fix it on a separate PR, or even disable the `move to space`  option int he UI branch 

If I move a chart in dashboard into another space (which doesn't make sense anyway) , the chart becomes "visible" 

![image](https://github.com/lightdash/lightdash/assets/1983672/ed158e3f-d5d9-49d4-8196-83831bc3582e)

ZeRego commented:
> If I move a chart in dashboard into another space (which doesn't make sense anyway) , the chart becomes "visible"

We will have the option to move charts from dashboards to the space. The BE will set the dashboard to null in that case. In your case we set the space uuid but didn't set the dashboarduuid to null causing that problem.
